File: practice/projects/expense-app-manager-java/src/main/java/com/alvaro8317/expense/handler/SpendHandler.java
package com.alvaro8317.expense.handler;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.alvaro8317.expense.model.Spend;
import com.alvaro8317.expense.repository.DynamoDbRepoSpend;
import com.alvaro8317.expense.service.SpendService;
import com.fasterxml.jackson.databind.ObjectMapper;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;

import java.math.BigDecimal;
import java.util.Map;

public class SpendHandler implements RequestHandler<Map<String, Object>, Map<String, Object>> {
    private final ObjectMapper mapper = new ObjectMapper();
    private final SpendService service;
    private final String tableName;

    public SpendHandler() {
        String tn = System.getenv("TABLE_NAME");
        DynamoDbClient client = DynamoDbClient.create();
        this.tableName = tn;
        this.service = new SpendService(new DynamoDbRepoSpend(client, tn));
    }

    public SpendHandler(SpendService service, String tableName) {
        this.service = service;
        this.tableName = tableName;
    }

    @Override
    @SuppressWarnings("unchecked")
    public Map<String, Object> handleRequest(Map<String, Object> event, Context context) {
        try {
            String body = (String) event.get("body");
            Map<String, String> payload = mapper.readValue(
                body, new com.fasterxml.jackson.core.type.TypeReference<Map<String, String>>() {}
            );

            Spend created = service.createSpend(
                payload.get("name"),
                payload.get("description"),
                new BigDecimal(payload.get("amount"))
            );

            return Map.of(
                "statusCode", 200,
                "headers", Map.of("Content-Type", "application/json"),
                "body", mapper.writeValueAsString(Map.of(
                    "message", "Spend created successfully",
                    "spend_created", created
                ))
            );
        } catch (Exception e) {
            return Map.of(
                "statusCode", 500,
                "headers", Map.of("Content-Type", "application/json"),
                "body", "{\"error\": \"Internal server error\"}"
            );
        }
    }
}

File: practice/projects/expense-app-manager-ts/src/handlers/spendHandler.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from "aws-lambda"
import { Spend } from "../models/spend"
import { DynamoDbRepoSpend } from "../repositories/dynamoDbRepoSpend"
import { SpendService } from "../services/spendService"

export async function spendHandler(
    event: APIGatewayProxyEvent,
    context: Context
): Promise<APIGatewayProxyResult> {
    try {
        const body = event.body ? JSON.parse(event.body) : {}
        const repo = new DynamoDbRepoSpend()
        const service = new SpendService(repo)

        const spendCreated: Spend = await service.createSpend(
            body.name,
            body.description,
            Number(body.amount)
        )

        return {
            statusCode: 200,
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                message: "Spend created successfully",
                spend_created: spendCreated.asDict(),
            }),
        }
    } catch (error) {
        console.error("Error creating spend:", error)
        return {
            statusCode: 500,
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ message: "Internal server error" }),
        }
    }
}

export async function getSpendHandler(
    event: APIGatewayProxyEvent,
    context: Context
): Promise<APIGatewayProxyResult> {
    try {
        const repo = new DynamoDbRepoSpend()
        const service = new SpendService(repo)

        const spends = await service.getSpends()

        return {
            statusCode: 200,
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ spends }),
        }
    } catch (error) {
        console.error("Error fetching spends:", error)
        return {
            statusCode: 500,
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ message: "Internal server error" }),
        }
    }
}
